File: python/tvm/runtime/executor/aot_executor.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""A Python wrapper for the Module-based Model Runtime Interface for Ahead-of-Time compilation."""

import numpy as np


class AotModule(object):
    """Wraps the AOT executor runtime.Module.

    This is a thin wrapper of the underlying TVM module.
    you can also directly call set_input, run, and get_output
    of underlying module functions

    Parameters
    ----------
    module : tvm.runtime.Module
        The internal tvm module that holds the implemented model functions.

    Attributes
    ----------
    module : tvm.runtime.Module
        The internal tvm module that holds the implemented model functions.

    Examples
    --------

    .. code-block:: python

        import tvm
        from tvm import relay
        from tvm.contrib import graph_executor

        # build the library using graph executor
        lib = relay.build(...)
        lib.export_library("compiled_lib.so")
        # load it back as a runtime
        lib: tvm.runtime.Module = tvm.runtime.load_module("compiled_lib.so")
        # Call the library factory function for default and create
        # a new runtime.Module, wrap with aot module.
        gmod = tvm.runtime.executor.AotModule(lib["default"](dev))
        # use the aot  module.
        gmod.set_input("x", data)
        gmod.run()
    """

    def __init__(self, module):
        self.module = module
        self._set_input = module["set_input"]
        self._run = module["run"]
        self._get_output = module["get_output"]
        self._get_input = module["get_input"]
        self._get_num_outputs = module["get_num_outputs"]
        self._get_input_index = module["get_input_index"]
        self._get_num_inputs = module["get_num_inputs"]

    def set_input(self, key=None, value=None, **params):
        """Set inputs to the module via kwargs

        Parameters
        ----------
        key : int or str
           The input key

        value : the input value.
           The input key

        params : dict of str to NDArray
           Additional arguments
        """
        if key is not None:
            v = self._get_input(key)
            if v is None:
                raise RuntimeError("Could not find '%s' in model's inputs" % key)
            v.copyfrom(value)

        if params:
            # upload big arrays first to avoid memory issue in rpc mode
            keys = list(params.keys())
            keys.sort(key=lambda x: -np.prod(params[x].shape))
            for k in keys:
                # TODO(zhiics) Skip the weights for submodule in a better way.
                # We should use MetadataModule for initialization and remove
                # params from set_input
                val = self._get_input(k)
                if val:
                    self._get_input(k).copyfrom(params[k])

    def run(self, **input_dict):
        """Run forward execution of the model

        Parameters
        ----------
        input_dict: dict of str to NDArray
            List of input values to be feed to
        """
        if input_dict:
            self.set_input(**input_dict)
        self._run()

    def get_num_outputs(self):
        """Get the number of outputs from the model

        Returns
        -------
        count : int
            The number of outputs.
        """
        return self._get_num_outputs()

    def get_num_inputs(self):
        """Get the number of inputs to the model

        Returns
        -------
        count : int
            The number of inputs.
        """
        return self._get_num_inputs()

    def get_input(self, index, out=None):
        """Get index-th input to out

        Parameters
        ----------
        index : int
            The input index

        out : NDArray
            The output array container
        """
        if out:
            self._get_input(index).copyto(out)
            return out

        return self._get_input(index)

    def get_input_index(self, name):
        """Get inputs index via input name.

        Parameters
        ----------
        name : str
           The input key name

        Returns
        -------
        index: int
            The input index. -1 will be returned if the given input name is not found.
        """
        return self._get_input_index(name)

    def get_output(self, index, out=None):
        """Get index-th output to out

        Parameters
        ----------
        index : int
            The output index

        out : NDArray
            The output array container
        """
        if out:
            self._get_output(index, out)
            return out

        return self._get_output(index)

File: src/runtime/aot_executor/aot_executor.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \brief Defines an implementation of Module-based Model Runtime Interface that works with
 *        Ahead-of-Time compilation.
 * \file aot_executor.cc
 */

#include "aot_executor.h"

#include <tvm/runtime/c_runtime_api.h>
#include <tvm/runtime/data_type.h>
#include <tvm/runtime/name_transforms.h>

#include <limits>
#include <memory>

#include "../meta_data.h"

namespace tvm {
namespace runtime {

AotExecutor::AotExecutor(tvm::runtime::Module module, const std::vector<Device>& devs)
    : module_{module}, devices_{devs} {
  auto fmetadata = module->GetFunction("get_metadata");
  CHECK(fmetadata != nullptr) << "Expected a module with PackedFunc get_metadata";
  auto ret_value = fmetadata();
  metadata_ = ret_value.AsObjectRef<tvm::runtime::metadata::Metadata>();

  ICHECK_EQ(devices_.size(), 1) << "Expect exactly 1 device passed.";
  DLDevice expected_device{kDLCPU, 0};
  ICHECK_EQ(devices_[0].device_id, expected_device.device_id)
      << "At this time, AOTExecutor supports only execution on kDLCPU 0";
  // TODO(tvm-team): Temporary hack since Hexagon is defined different than kDLCPU.
  bool is_valid_device =
      (devices_[0].device_type == kDLHexagon) || (devices_[0].device_type == kDLCPU);
  CHECK(is_valid_device)
      << "At this time, AOTExecutor supports only execution on kDLCPU 0 or kDLHexagon 0";

  for (auto input : metadata_->inputs()) {
    // TODO(areusch): Encode device information in Metadata.
    args_.emplace_back(NDArray::Empty(ShapeTuple(input->shape().begin(), input->shape().end()),
                                      input->dtype(), devices_[0]));
  }

  for (auto output : metadata_->outputs()) {
    args_.emplace_back(NDArray::Empty(ShapeTuple(output->shape().begin(), output->shape().end()),
                                      output->dtype(), devices_[0]));
  }

  // USMP is used
  if (metadata_->num_workspace_pools()) {
    // merge all constants into one ndarray
    int64_t blob_len = 0;
    for (const auto& c : metadata_->constant_pools()) {
      auto data = c->data();
      int64_t byte_size = GetDataSize(*data.operator->()) + c->byte_offset();
      blob_len = blob_len > byte_size ? blob_len : byte_size;
    }
    ICHECK(blob_len < std::numeric_limits<int32_t>::max());
    NDArray ci = NDArray::Empty({blob_len}, DataType::UInt(8), devices_[0]);
    for (const auto& c : metadata_->constant_pools()) {
      auto data = c->data();
      data.CopyToBytes(static_cast<uint8_t*>(ci->data) + c->byte_offset(),
                       GetDataSize(*data.operator->()));
    }
    // Emplace constant node pool only if workspace pools supplied
    args_.emplace_back(ci);

    int32_t pool_len = 0;
    for (auto pool : metadata_->workspace_pools()) {
      pool_len =
          GetDataSize(*NDArray::Empty({pool->shape()}, pool->dtype(), devices_[0]).operator->());
      args_.emplace_back(NDArray::Empty({pool_len}, DataType::UInt(8), devices_[0]));
    }
  }
}

PackedFunc AotExecutor::GetFunction(const String& name, const ObjectPtr<Object>& sptr_to_self) {
  // Return member functions during query.
  if (name == "set_input") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0])) {
        int in_idx = this->GetInputIndex(tvm::runtime::SanitizeName(args[0].operator String()));
        if (in_idx >= 0) this->SetInput(in_idx, args[1]);
      } else {
        this->SetInput(args[0], args[1]);
      }
    });
  } else if (name == "set_input_zero_copy") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0])) {
        int in_idx = this->GetInputIndex(tvm::runtime::SanitizeName(args[0].operator String()));
        if (in_idx >= 0) this->SetInputZeroCopy(in_idx, args[1]);
      } else {
        this->SetInputZeroCopy(args[0], args[1]);
      }
    });
  } else if (name == "set_output_zero_copy") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (String::CanConvertFrom(args[0])) {
        int out_idx = this->GetOutputIndex(tvm::runtime::SanitizeName(args[0].operator String()));
        if (out_idx >= 0) this->SetOutputZeroCopy(out_idx, args[1]);
      } else {
        this->SetOutputZeroCopy(args[0], args[1]);
      }
    });
  } else if (name == "get_output") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      if (args.num_args == 2) {
        this->CopyOutputTo(args[0], args[1]);
      } else {
        *rv = this->GetOutput(args[0]);
      }
    });
  } else if (name == "get_input") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      int in_idx = 0;
      if (String::CanConvertFrom(args[0])) {
        in_idx = this->GetInputIndex(tvm::runtime::SanitizeName(args[0].operator String()));
      } else {
        in_idx = args[0];
      }
      if (in_idx >= 0) {
        *rv = this->GetInput(in_idx);
      }
    });
  } else if (name == "get_num_outputs") {
    return PackedFunc(
        [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->NumOutputs(); });
  } else if (name == "get_num_inputs") {
    return PackedFunc(
        [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->NumInputs(); });
  } else if (name == "run") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { this->Run(); });
  } else if (name == "get_input_index") {
    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
      CHECK(String::CanConvertFrom(args[0])) << "Input key is not a string";
      *rv = this->GetInputIndex(tvm::runtime::SanitizeName(args[0].operator String()));
    });
  } else if (name == "get_input_name") {
    return PackedFunc(
        [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->GetInputName(args[0]); });
  } else {
    return PackedFunc();
  }
}

void AotExecutor::Run() {
  auto pf = module_.GetFunction(
      get_name_mangled(metadata_->mod_name(), ::tvm::runtime::symbol::tvm_module_main),
      true /* query_imports */);
  ICHECK(pf != nullptr) << "Module entrypoint is not defined";

  const int num_args = args_.size();
  auto call_values = ::std::make_unique<TVMValue[]>(num_args);
  auto call_type_codes = ::std::make_unique<int[]>(num_args);
  for (int i = 0; i < num_args; ++i) {
    auto managed = args_[i].ToDLPack();
    call_values.get()[i].v_handle = &managed->dl_tensor;
    call_type_codes.get()[i] = kTVMDLTensorHandle;
  }

  TVMArgs args{call_values.get(), call_type_codes.get(), num_args};
  TVMRetValue rv;
  pf.CallPacked(args, &rv);
}

int AotExecutor::GetInputIndex(const std::string& name) {
  auto inputs = metadata_->inputs();
  for (unsigned int i = 0; i < inputs.size(); i++) {
    if (inputs[i]->name() == name) {
      return i;
    }
  }
  ICHECK(false) << "Invalid input name.";
}

std::string AotExecutor::GetInputName(int index) {
  auto inputs = metadata_->inputs();
  return inputs[index]->name();
}

int AotExecutor::GetOutputIndex(const std::string& name) {
  auto outputs = metadata_->outputs();
  for (unsigned int i = 0; i < outputs.size(); i++) {
    if (outputs[i]->name() == name) {
      return i;
    }
  }
  return -1;
}

void AotExecutor::SetInput(int index, DLTensor* data_ref) { args_[index].CopyFrom(data_ref); }

void AotExecutor::SetInputZeroCopy(int index, DLTensor* data_ref) {
  ICHECK(false) << "not implemented";
}

void AotExecutor::SetOutputZeroCopy(int index, DLTensor* data_ref) {
  ICHECK(false) << "not implemented";
}

int AotExecutor::NumOutputs() const { return metadata_->num_outputs(); }

int AotExecutor::NumInputs() const { return metadata_->num_inputs(); }

NDArray AotExecutor::GetInput(int index) const { return args_[index]; }

NDArray AotExecutor::GetOutput(int index) const { return args_[metadata_->num_inputs() + index]; }

void AotExecutor::CopyOutputTo(int index, DLTensor* data_out) { GetOutput(index).CopyTo(data_out); }

}  // namespace runtime
}  // namespace tvm
